### DISCUSSION TITLE
Reference Resolution in ast.Node

### DISCUSSION THREAD

**Discussion by dcb1001** [cue-lang/cue#3171] [Q&A]
### What version of CUE are you using (`cue version`)?

<pre>
$ cue version
cue version v0.7.0
</pre>

### What did you do?
I am trying to parse a cue file using the `cue.Iterator`. However, the `cue.Value` doesn't store the if-else block and for loops. Therefore, I am trying to parse it by converting `cue.Value` to `ast.Node`, but facing issues regarding reference resolution.
```
-- cue.mod/pkg/example.com/schema/schema.cue --

package schema 

body: {
    a: int
    b: "hello"
}

-- demo.cue --
package config 

import (
    "tower.org/schema"
)
a: int
d: float
body: {

    c: schema.body.a
    d: int
    e: c
    if c != null {
        b: schema.body.a
        b2: d | *3

        if c == 10 {
            b3: schema.body.a
            b4: b2
            b5: a
        }
    }
}

-- main.go --
package main

import (
	"fmt"
	"log"
	"os"

	"cuelang.org/go/cue"
        "cuelang.org/go/cue/token"
        "cuelang.org/go/cue/ast"
        "cuelang.org/go/cue/ast/astutil"
	"cuelang.org/go/cue/cuecontext"
	"cuelang.org/go/cue/load"
)

func helper2 (node ast.Node, ctx *cue.Context, data cue.Value) {
    switch n := node.(type) {
        case *ast.StructLit: 
            for _, ele := range n.Elts{
                helper2(ele, ctx, data)
            }
        case *ast.Field:
            fmt.Println("\nfield ast label:", n.Label)
            fmt.Println("field ast value :", n.Value)

            astutil.ResolveExpr(n.Value, func (pos token.Pos, msg string, args ...interface{}){})
            value := ctx.BuildExpr(n.Value, []cue.BuildOption{
                cue.Scope(data),
            }...)
            fmt.Println("field value :", value)
        case *ast.Comprehension:
            fmt.Println("\nthis is comprehension\n")
            helper2(n.Value, ctx, data)
            

    }
}

func helper (data cue.Value, original cue.Value, opts []cue.Option, ctx *cue.Context) {
    fieldItr, err := data.Fields(opts...)
    if err != nil {
        log.Fatal(err)
    }

    for fieldItr.Next() {
        fieldName := fieldItr.Selector().String()
        fmt.Println("\nfieldName", fieldName)

        valueItr, err := fieldItr.Value().Fields(opts...)
        if err != nil {
            log.Println(err)
        }

        if valueItr.Next() == false {
            fmt.Println("leafvalue:", fieldItr.Value())
        }else {
            node := fieldItr.Value().Syntax([]cue.Option{
                cue.All(),
                cue.DisallowCycles(true),
                cue.InlineImports(true),
                cue.Optional(true),
                cue.Raw(),
                cue.Concrete(true),
            }...)
            helper2(node, ctx, data)

        }


    }
}

func main () {
    if len(os.Args) < 2 {
        fmt.Println("usage : go run hello2.go [cuefile path]")
        return
    }

    config := os.Args[1]

    ctx := cuecontext.New()
    args := []string{
		config,
	}

	binst := load.Instances(args, nil)
	insts, instsErr := ctx.BuildInstances(binst)

	if instsErr != nil {
		log.Fatal(instsErr)
	}

	if len(insts) != 1 {
        log.Fatal("more than one instance")
    }

    opts := []cue.Option {
        cue.All(),
    }

    
   helper(insts[0], insts[0], opts, ctx)

}
```

### What did you expect to see?
Everything resolved to their respective references

### What did you see instead?
```
> go run hello2.go ./demo.cue

fieldName a
leafvalue: int

fieldName d
leafvalue: float

fieldName body

field ast label: c
field ast value : A
field value : _|_ // reference "A" not found

field ast label: d
field ast value : int
field value : int

field ast label: e
field ast value : c
field value : _|_ // reference "c" not found

this is comprehension


field ast label: b
field ast value : A
field value : _|_ // reference "A" not found

field ast label: b2
field ast value : &{d - | 0xc0002b0b40 {<nil>} {{}}}
field value : *3 | float

this is comprehension


field ast label: b3
field ast value : A
field value : _|_ // reference "A" not found

field ast label: b4
field ast value : b2
field value : _|_ // reference "b2" not found

field ast label: b5
field ast value : int
field value : int
```

**Comment by myitcv** [comment]
Hi @dcb1001 - I've converted your issue to a discussion for now. Happy to create an issue if indeed there is a bug here.

> I am trying to parse a cue file using the cue.Iterator. However, the cue.Value doesn't store the if-else block and for loops. Therefore, I am trying to parse it by converting cue.Value to ast.Node, but facing issues regarding reference resolution.

Stepping back, can you explain what you are trying to do, and why you need to be able to address the `if` and `for` comprehensions? That will help to guide the approach and hence the answer. Thanks

**Reply by dcb1001** [reply]
Hey @myitcv ,
I have a use case where I have to define a schema for the application in cue. However, the constraint is that the config is provided by the user which is validated against the schema and used to configure the application. 
Now, these applications are written in C++ and since there are no APIs to support CUE in C++, I am parsing the CUE using Go and building my custom tree which is passed to C++ runtime. 

Now, assuming app dev wants to express a schema
```
a ?: int
b ?: int
if a == 10 || a == 20 {
    b !: int
}
```
Since, I haven't provided the data, these `if` expressions aren't resolved while building the cue file using `cue.BuildInstances` and it raises the error `cannot reference optional field: a:
    -:3:4
cannot reference optional field: a:
    -:3:15
`
which is suppressed by me, given if the data was provided, it would have been a valid schema-config pair.

```Note : - We can't use cgo to provide an interface between C++ and Go, because of concerns regarding Go GC, signal handling and spawning of threads by Go runtime (primarily gc.Scavenge, gc.Marker, gc.Sweep)```

Given, these constraints, I want to support maximum syntax of cue. For eg: -

```
-- Possibility 1 --
a : int
b ?: int
if a == 10 || a == 20 {
    b !: int
}

-- Possibility 2 --
a : int 
b ?: int
if a == 10 {
    b !: int
}

if a == 20 {
    b !: int
}
```

Since, the `config` will be available at the runtime of the application and I can't involve Go runtime at the runtime of application. I am trying to parse the cue file at build time using Go which dumps the tree that is used by C++ application to configure the application. I don't want to write my own cue parser and use CUE APIs as much as possible, so I was thinking of using the CUE AST to get the `if condition` and get the `reference resolved if's body` from the `ast.Node` to `cue.Value` which can easily be iterated over using `*cue.Iterator`.

**Reply by myitcv** [reply]
> Now, assuming app dev wants to express a schema

That would be written as follows in CUE:

```
a?: int
b?: int
if a != _|_ if a == 10 || a == 20 {
	b!: int
}
```

(noting that #943 has some builtins that will make it much more precise and readable what the comparison to `_|_` represents).


> We can't use cgo

So just to double-check you can't or don't want to use https://github.com/cue-lang/libcue? 

> Given, these constraints, I want to support maximum syntax of cue

CUE's APIs exist at two "levels": the [AST](https://pkg.go.dev/cuelang.org/go/cue/ast) and the [`cue.Value`](https://pkg.go.dev/cuelang.org/go/cue). The former is the result of a straight parse of CUE, the latter is the result of evaluation (i.e. the combining of all the constraints for given fields etc). What I understand from your situation is that you're roughly looking to export the evaluated result from, say, the CUE above (more precisely some subset of CUE), into a structured form, and then write some C++ to interpret that structured form in a way similar to the evaluator. 

Is this correct?

If so, we do not currently expose the APIs to allow for this. Nor am I clear that this is something we will prioritise in the short term. Also, this would represent a fairly significant amount of effort on your side, to implement a limited form of the evaluator. Which is why I would encourage exploring whether `libcue` can do what you want. It is in an early form right now, but is forming the basis of work we are doing with [Java, Python and other languages](https://github.com/orgs/cue-lang/projects/24). 

cc @4ad

**Reply by dcb1001** [reply]
> CUE's APIs exist at two "levels": the [AST](https://pkg.go.dev/cuelang.org/go/cue/ast) and the [cue.Value](https://pkg.go.dev/cuelang.org/go/cue). The former is the result of a straight parse of CUE, the latter is the result of evaluation (i.e. the combining of all the constraints for given fields etc). What I understand from your situation is that you're roughly looking to export the evaluated result from, say, the CUE above (more precisely some subset of CUE), into a structured form, and then write some C++ to interpret that structured form in a way similar to the evaluator.

This is partially correct. **The Go implements the parser and evaluator for CUE and builds custom AST. Afterwards, it dump the custom AST in form of adjacency list which is consumed by C++ to rebuild the tree and perform other operations.**

> So just to double-check you can't or don't want to use https://github.com/cue-lang/libcue?

The reason of not using `cgo` is that the C++ applications are single-threaded and run in a very controlled environment with limited resources. 
**Problem 1: -**
After invoking the Go runtime, the threads spawned and managed by Go runtime remain live until the process terminates. Also, go scheduler might cache some threads, in hope of executing a Go function in the future during the lifetime of the process. This optimization leads to unnecessary live threads in process memory.

**Problem 2: -**
Another is the issue of Go GC, the C++ applications are pinned to a specific core and there might be a case when GOGC threads does "stop the world" which would lead to the blockage of business logic.

and many more.

An off-topic question, is there a way to close the GO runtime without terminating the process? 

Back to CUE, I tried some work-around to resolve the references of the subset of CUE (for eg: - body of IfClause) https://github.com/cue-lang/cue/discussions/3125
Since, cue treats `{}` as `ast.Structlit`. It would be helpful if I could build `ast.Comprehenion`'s `value` which is `cue.Structlit` into `cue.Value` which is easy to iterate over. Is there a way to do that?  
Because, resolving the body of `if` statement is basically writing my own cue parser and I don't want to do that if I don't have to. Since, handling flexibility of syntax, reference resolution, Alias handling, cycle detection etc. has to be done.

**Reply by myitcv** [reply]
Understood. CUE and its various APIs do not expose the information you are going to need for this AFAICT. Effectively what you are looking for is an encoding of the post-evaluation encoding of CUE so that an alternative (limited functionality) evaluator (your C++ implementation) can interpret it. The reason I think you would need at least a minimal evaluator is that the CUE above is incomplete/partial: there are no values for `a` or `b`. Hence the encoding would need to handle taking a values and combining with the incomplete CUE... which is the most basic primitive of evaluation. 

I will leave @4ad to answer questions on the `libcue` side of things.

**Reply by 4ad** [reply]
`libcue` just exposes the high-level CUE API, so even if @dcb1001 could use `libcue`, it wouldn't solve his problem.

Ultimately, @myitcv is right. CUE is always partially evaluated which is why you can always add more configuration to unify with later. If you want to further process the result of an incomplete CUE evaluation in C++, you need to write a C++ CUE implementation.

**Reply by dcb1001** [reply]
Thanks @myitcv  and @4ad  for the input !!
